include/dali/core/geom/box.h

```diff
@@ -0,0 +1,148 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_CORE_GEOM_BOX_H_
+#define DALI_CORE_GEOM_BOX_H_
+
+#include "dali/core/geom/vec.h"
+
+namespace dali {
+
+template<size_t ndims, typename CoordinateType>
+struct Box {
+  static_assert(std::is_standard_layout<CoordinateType>::value,
+                "Cannot create a Box of a non-standard layout type");
+  using corner_t = vec<ndims, CoordinateType>;
+  static_assert(std::is_pod<corner_t>::value, "Corner has to be POD");
+
+  /**
+   * Corners of the box.
+   * Assumes, that `lo <= hi`, i.e. every coordinate of `lo` will be lower or equal to
+   * corresponding coordinate of `hi`.
+   */
+  corner_t lo, hi;
+
+
+  /**
+   * Creates an empty box
+   */
+  Box() = default;
+
+
+  /**
+   * Creates a bounding-box using 2 points.
+   * Assumes, that `lo <= hi`, i.e. every coordinate of `lo` will be lower or equal to
+   * corresponding coordinate of `hi`.
+   */
+  Box(const corner_t &lo, const corner_t &hi) {
+    this->lo = lo;
+    this->hi = hi;
+  }
+
+
+  /**
+   * Convenient overload, for 2-D case
+   * @param left x-coordinate of `lo`
+   * @param top y-coordinate of `lo`
+   * @param right x-coordinate of `hi`
+   * @param bottom y-coordinate of `hi`
+   */
+  Box(CoordinateType left, CoordinateType top, CoordinateType right, CoordinateType bottom) {
```

Comment:
That's wrong. It can't work for `ndims != 2`.

Reply:
Still, most of our data uses 2-dim boxes, as we work with images.

I think that splitting that into specializations per ndim would be an overkill, so I would probably use something like:

template <typename... Lo, typename... Hi>
Box(Lo... l, Hi... h) : lo{l...}, hi{h...} {
static_assert(sizeof...(Lo) == ndim && sizeof...(Hi) == ndim);
}

How to actually divide those argument packs I would leave as an exercise to the reader.